src/sysen2/mldev.108

```diff
@@ -94,69 +96,95 @@ GO:	MOVE P,[-LPDLL-1,,PDL-1]
+		movem c,jbcdev	; store as remote device
+		move a,b	; get host in A for gothst
+		jrst gothst ]	; and go process request for this host
+
+nox:				; jump here if hostname minus first "x" is not a known ITS
+				; check to see if format is hhDIR
 	move c,a
-	lsh c,12.
-	cain c,'DIR
-	 jrst [ ldb b,[.bp (777700),a] ;get xx part in b
-		pushj p,isitsp
-		 jrst .+1
-		move a,b	; get host in A
-		jrst gothst ]		
-	move b,a
-	;; is b an ITS name?
-	pushj p,isitsp
-	 jrst notits
-	  jrst gothst
+	lsh c,12.		; move past hostname
+	camn c,dirdev		; is what remains the DIR device?
+	  jrst dirchk		;  yes, continue checking hhDIR case
+
+				; is it hhhdir or hhh
+	hrlz c,a		; get device in c
+	jumpe c, host3		; jump if no device
+
+	ldb c,[000600,,a]	; can only be hhddd if last sixbit character is null
+	jumpe c,hhddd1		; handle hhddd case if last character is NUL
 	
-notits:	;; nothing came of it, fall through
-	MOVS A,JBCDEV
-	CAIN A,'XGP	;XGP THROUGH THE NET TO AI FROM ML/MC/DM
-	 JRST [	MOVS A,[SIXBIT/AIXGP/]
-		MOVSM A,JBCDEV
-		MOVEI A,'AI
-		JRST GOTHST]
-	CAIN A,'GLP		;GOULD THROUGH THE NET TO MX
-	 JRST [	MOVS A,[SIXBIT/MXGLP/]
-		MOVSM A,JBCDEV
-		MOVEI A,'MX
-		JRST GOTHST]
-	CAIN A,'DVR	;DVR THROUGH THE NET TO MC FROM AI/ML/DM
-	 JRST [	MOVS A,[SIXBIT/MCDVR/]
-		MOVSM A,JBCDEV
-		MOVEI A,'MC
-		JRST GOTHST]
-	CAIN A,'DVS	;DVS THROUGH THE NET TO MC FROM ML/DM
-	 JRST [	MOVS A,[SIXBIT/MCDVS/]
-		MOVSM A,JBCDEV
-		MOVEI A,'MC
-		JRST GOTHST]
-	CAIN A,'TPL	;TPL THROUGH THE NETWORK FOR MC
-	 JRST [	MOVS A,[SIXBIT/MLTPL/]
-		MOVSM A,JBCDEV
-		LDB B,[000400,,JBCOP]
+	hrlz c,a		; check for hhhddd, get ddd in c
+	jumpn e,hhhddd		; is there a ddd?, if so handle it
+	jrst chkits		; nope, assume that we just have hhh
+
+host3:				; seems like hhh, but may be special, check to see if ITS
+	move b,a		; hostname in A
+	pushj p,isitsp		; is it a known ITS
+	  jrst notits		;  nope, see if special device
+	move c,dskdev		; we have an ITS name, so assume DSK device
+	movem c,jbcdev		; save remote device
+	jrst gothst		; and process our host
+
+dirchk:				; 2-char hostname with DIR device
+	ldb b,[.bp (777700),a]	; get hh part in b
+	lsh b,24.		; left justify hostname
+dirch2:	move c,dirdev		; set remote device as DIR device
+	movem c,jbcdev
+	move a,b		; get host in a
+	jrst gothst		; process host
+
+hhddd2: ldb a,[301400,,jbcdev]	; get hostname in A
+	lsh a,24.		; left justify hostname
+	movem c,jbcdev		; set remote device from C
+	jrst gothst		; process host
+
+hhddd1:				; continue check for hhddd
+	ldb c,[003000,,a]	; get device portion
+	lsh c,12.		; left-justify
+	jumpn c,hhddd2		; jump if we have a device; get hostname and set remote device
+
+	;; fall through; assume just an ITS name
+chkits:	move b,a
+	pushj p,isitsp		; is b an ITS name?
+	 jrst notits		;  no, process as special device
+        jrst gothst		; yes, process this host
+
+notits:				; here we've determined that hostname is invalid, check for special device
+	MOVS C,JBCDEV
```

Comment:
Since these machines, in these configurations, no longer exist and probably won't, I suggest commenting out this hackery.

Reply:
Well, I thought about this -- however it is good documentation as to where these additional devices used to live. Also, we may implement one or more of these devices at some point (TPL, XGP), and it may be that we can only implement them on a KA (or KL or KS) and if that is the case, this logic might still make some sense.  I don't see what the harm is in leaving it.

Reply:
I like to entertain the idea that some day the classic AI, DM, ML, MC machines be brought back under simulation.  I also think think they are have an important place in ITS history and should stay in ITS source code files; I believe this is in the spirit of ITS hacking.